csvcubed/tests/behaviour/steps/qbwriter.py

```diff
@@ -735,3 +743,89 @@ def step_impl(context, cube_name: str):
 def step_impl(context):
     rdf_to_add = context.text
     context.turtle += rdf_to_add
+
+
+@Then("the cube's metadata should contain URLs with file endings")
+def step_impl(context):
+    temp_dir = get_context_temp_dir_path(context)
+    assertURIStyle(URIStyle.Standard, temp_dir, context.csv_file_name)
+
+
+@Then("the cube's metadata should contain URLs without file endings")
+def step_impl(context):
+    temp_dir = get_context_temp_dir_path(context)
+    assertURIStyle(URIStyle.WithoutFileExtensions, temp_dir, context.csv_file_name)
+
+
+@Given(
+    'a single-measure QbCube named "{cube_name}" configured with "{uri_style}" URI style'
+)
+def step_impl(context, cube_name: str, uri_style: str):
+    context.cube = _get_single_measure_cube_with_name_and_id(
+        cube_name, None, URIStyle[uri_style]
+    )
+
+
+def assertURIStyle(uri_style: URIStyle, temp_dir: Path, csv_file_name: str):
+    metadataFilePath = temp_dir.joinpath(f"{csv_file_name}-metadata.json")
+    g = Graph()
+    g.parse(metadataFilePath)
+
+    predicate_whitelist = [
+        str(RDF.type),
+        "http://www.w3.org/ns/dcat#landingPage",
+        "http://purl.org/dc/terms/creator",
+        "http://purl.org/dc/terms/publisher",
+        "http://www.w3.org/ns/dcat#contactPoint",
+    ]
+    object_prefix_whitelist = [
+        str(XSD),
+        "http://gss-data.org.uk",
+        "http://www.w3.org",
+        "http://www.nationalarchives.gov.uk",
+        "http://purl.org",
+    ]
+    uri_literal_whitelist = ["rdf:type"]
+    uri_data_types = [
+        "http://www.w3.org/ns/csvw#uriTemplate",
+        "http://www.w3.org/2001/XMLSchema#anyURI",
+    ]
+
+    uriRefSubjects = set([s for s in g.subjects() if isinstance(s, URIRef)])
+    assertURIStyles(uri_style, uriRefSubjects)
+
+    uriRefObjects = set(
+        [
+            o
+            for (p, o) in g.predicate_objects()
+            if isinstance(o, URIRef)
+            and not (str(p) in predicate_whitelist)
+            and not (str(o).startswith(tuple(object_prefix_whitelist)))
+        ]
+    )
+    assertURIStyles(uri_style, uriRefObjects)
+
+    uriLiteralObjects = set(
```

Comment:
You can do set comprehensions just like list comprehensions in python, i.e. switch the `[ ... ]` for `{ ... }` and you don't need to convert to a set.

Reply:
ooo nice